### PR TITLE
Refactored the codegen into smaller functions, in anticipation of extensions.

### PR DESCRIPTION
**Pull request by haberman**
This PR contains no behavior change.

### CODE REVIEW

File: upbc/protoc-gen-upb.cc [hunk 1]
```diff
@@ -554,7 +554,7 @@ void WriteHeader(const protobuf::FileDescriptor* file, Output& output) {
   // Order by full name for consistent ordering.
   std::map<std::string, const protobuf::Descriptor*> forward_messages;
 
-  for (auto message : SortedMessages(file)) {
+  for (auto message : this_file_messages) {
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:557]: auto*, don't make me worry about copies :P

File: upbc/protoc-gen-upb.cc [hunk 2]
```diff
@@ -843,139 +843,163 @@ std::vector<TableEntry> FastDecodeTable(const protobuf::Descriptor* message,
...
+  output("  ");
+  WriteField(field, GetSizeInit(layout.GetFieldOffset(field)), presence,
+             submsg_index, output);
+  output(",\n");
+}
 
+// Writes a single message into a .upb.c source file.
+void WriteMessage(const protobuf::Descriptor* message, Output& output,
+                  bool fasttable_enabled) {
+  std::string msgname = ToCIdent(message->full_name());
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:894]: msg_name

File: upbc/protoc-gen-upb.cc [hunk 3]
```diff
@@ -843,139 +843,163 @@ std::vector<TableEntry> FastDecodeTable(const protobuf::Descriptor* message,
...
+  std::string fields_array_ref = "NULL";
+  std::string submsgs_array_ref = "NULL";
+  uint8_t dense_below = 0;
+  int dense_below_max = std::numeric_limits<decltype(dense_below)>::max();
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:898]: const?

File: upbc/protoc-gen-upb.cc [hunk 4]
```diff
@@ -843,139 +843,163 @@ std::vector<TableEntry> FastDecodeTable(const protobuf::Descriptor* message,
...
-      "\n");
+  output("  ");
+  WriteField(field, GetSizeInit(layout.GetFieldOffset(field)), presence,
+             submsg_index, output);
+  output(",\n");
+}
 
+// Writes a single message into a .upb.c source file.
+void WriteMessage(const protobuf::Descriptor* message, Output& output,
+                  bool fasttable_enabled) {
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:893]: fast_table_enabled

File: upbc/protoc-gen-upb.cc [hunk 5]
```diff
@@ -843,139 +843,163 @@ std::vector<TableEntry> FastDecodeTable(const protobuf::Descriptor* message,
...
+    output("  }),\n");
+  }
+  output("};\n\n");
+}
 
-    std::vector<TableEntry> table;
-    uint8_t table_mask = -1;
+void WriteMessages(const protobuf::FileDescriptor* file, Output& output,
+                   bool fasttable_enabled) {
+  for (auto message : SortedMessages(file)) {
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:976]: auto* ?

File: upbc/protoc-gen-upb.cc [hunk 6]
```diff
@@ -843,139 +843,163 @@ std::vector<TableEntry> FastDecodeTable(const protobuf::Descriptor* message,
...
+    WriteMessage(message, output, fasttable_enabled);
+  }
+}
 
-    if (fasttable_enabled) {
-      table = FastDecodeTable(message, layout);
-    }
+// Writes a .upb.c source file.
+void WriteSource(const protobuf::FileDescriptor* file, Output& output,
+                 bool fasttable_enabled) {
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:983]: fast_table_enabled

File: upbc/protoc-gen-upb.cc (hunk 1 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:557]: Done.

File: upbc/protoc-gen-upb.cc (hunk 2 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:894]: Done.

File: upbc/protoc-gen-upb.cc (hunk 3 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:898]: Done.

**Review by fowles — APPROVED** [review]

File: upbc/protoc-gen-upb.cc (hunk 4 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:893]: I would prefer to keep this, it matches:

- https://github.com/protocolbuffers/upb/blob/master/BUILD#L51-L60
- https://github.com/protocolbuffers/upb/blob/master/upb/msg_internal.h#L58
- https://github.com/protocolbuffers/upb/blob/master/upb/msg_internal.h#L72

It sets it apart as a specific term of art in the code. I think `minitable` would be a good follow-up term of art for mini tables.

File: upbc/protoc-gen-upb.cc (hunk 5 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:976]: Done.

File: upbc/protoc-gen-upb.cc (hunk 6 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:983]: Ditto the above.

**Pull request merged**